--- a/net/minecraft/world/level/block/entity/FuelValues.java
+++ b/net/minecraft/world/level/block/entity/FuelValues.java
@@ -24,7 +_,10 @@
     }
 
     public boolean isFuel(ItemStack stack) {
-        return this.values.containsKey(stack.getItem());
+        boolean isFuel = this.values.containsKey(stack.getItem());
+        net.craftoriya.events.IsFuelEvent event = new net.craftoriya.events.IsFuelEvent(stack.asBukkitCopy(), isFuel);
+        org.bukkit.Bukkit.getPluginManager().callEvent(event);
+        return event.isFuel();
     }
 
     public SequencedSet<Item> fuelItems() {
@@ -32,7 +_,10 @@
     }
 
     public int burnDuration(ItemStack stack) {
-        return stack.isEmpty() ? 0 : this.values.getInt(stack.getItem());
+        int burnDuration = stack.isEmpty() ? 0 : this.values.getInt(stack.getItem());
+        net.craftoriya.events.BurnDurationEvent event = new net.craftoriya.events.BurnDurationEvent(stack.asBukkitCopy(), burnDuration);
+        org.bukkit.Bukkit.getPluginManager().callEvent(event);
+        return event.getBurnDuration();
     }
 
     public static FuelValues vanillaBurnTimes(HolderLookup.Provider registries, FeatureFlagSet enabledFeatures) {
